### PR TITLE
MNT: Remove jwst as dependency for Imviz

### LINKED ISSUE #725: Imviz: Remove jwst dependency
@jdavies-st said we don't need it. 🤷 

```python
import asdf

with asdf.open("jw001234_blah_blah_cal.fits") as af:
    wcs = af.tree["meta"]["wcs"]
```

### PR DESCRIPTION
**Pull request by pllim**
Fix #725 

<details>
<summary>This is no longer an issue.</summary>

But I am not convinced this works because I get this error locally. Test image is https://data.science.stsci.edu/redirect/JWST/jwst-data_analysis_tools/stellar_photometry/jw01072001001_01101_00001_nrcb1_cal.fits

```
...
gwcs: 0.16.1
asdf: 2.7.3
...
_____________________________________ TestParseImage.test_parse_jwst_nircam_level2 _____________________________________

    @pytest.mark.skipif(not HAS_JWST_ASDF, reason='asdf and gwcs not installed')
    @pytest.mark.remote_data
    def test_parse_jwst_nircam_level2(self, imviz_app):
        from gwcs import WCS as GWCS

        filename = download_file(self.jwst_asdf_url_1, cache=True)

        # Default behavior: Science image
>       parse_data(imviz_app.app, filename, show_in_viewer=False)

jdaviz/configs/imviz/tests/test_parser.py:193:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
jdaviz/configs/imviz/plugins/parsers.py:65: in parse_data
    _parse_image(app, pf, data_label, show_in_viewer, ext=ext)
jdaviz/configs/imviz/plugins/parsers.py:133: in _parse_image
    for data, data_label in data_iter:
jdaviz/configs/imviz/plugins/parsers.py:207: in _jwst_to_glue_data
    data, new_data_label = _jwst2data(file_obj, ext, data_label)
jdaviz/configs/imviz/plugins/parsers.py:219: in _jwst2data
    with asdf.open(file_obj) as af:
.../asdf/asdf.py:1480: in open_asdf
    return AsdfFile._open_impl(instance,
.../asdf/asdf.py:719: in _open_impl
    return fits_embed.AsdfInFits._open_impl(fd, uri=uri,
.../asdf/fits_embed.py:266: in _open_impl
    return cls._open_asdf(self, buff, uri=uri, mode='r',
.../asdf/asdf.py:693: in _open_asdf
    tree = yamlutil.tagged_tree_to_custom_tree(tree, self, _force_raw_types)
.../asdf/yamlutil.py:287: in tagged_tree_to_custom_tree
    return treeutil.walk_and_modify(
.../asdf/treeutil.py:406: in walk_and_modify
    return _recurse(top)
.../asdf/treeutil.py:387: in _recurse
    result = _handle_children(node, json_id)
.../asdf/treeutil.py:365: in _handle_children
    return _handle_generator(result)
.../asdf/treeutil.py:263: in _handle_generator
    result = next(generator)
.../asdf/treeutil.py:291: in _handle_mapping
    value = _recurse(value, json_id)
.../asdf/treeutil.py:387: in _recurse
    result = _handle_children(node, json_id)
.../asdf/treeutil.py:365: in _handle_children
    return _handle_generator(result)
.../asdf/treeutil.py:263: in _handle_generator
    result = next(generator)
.../asdf/treeutil.py:291: in _handle_mapping
    value = _recurse(value, json_id)
.../asdf/treeutil.py:388: in _recurse
    result = _handle_callback(result, json_id)
.../asdf/treeutil.py:270: in _handle_callback
    result = callback(node)
.../asdf/yamlutil.py:279: in walker
    return tag_type.from_tree_tagged(node, ctx)
.../asdf/types.py:407: in from_tree_tagged
    return cls.from_tree(tree.data, ctx)
.../gwcs/tags/wcs.py:29: in from_tree
    steps = [(x.frame, x.transform) for x in node['steps']]
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

.0 = <list_iterator object at ...>

>   steps = [(x.frame, x.transform) for x in node['steps']]
E   AttributeError: 'TaggedDict' object has no attribute 'frame'

.../gwcs/tags/wcs.py:29: AttributeError
```

</details>

cc @jdavies-st

### CODE REVIEW

**Comment by pllim** [comment]
Hmm! Different error when I upgrade to `asdf` 2.8.1. No upgrade is available for `gwcs`.

```
        # Default behavior: Science image
>       parse_data(imviz_app.app, filename, show_in_viewer=False)

jdaviz/configs/imviz/tests/test_parser.py:193:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
jdaviz/configs/imviz/plugins/parsers.py:65: in parse_data
    _parse_image(app, pf, data_label, show_in_viewer, ext=ext)
jdaviz/configs/imviz/plugins/parsers.py:133: in _parse_image
    for data, data_label in data_iter:
jdaviz/configs/imviz/plugins/parsers.py:207: in _jwst_to_glue_data
    data, new_data_label = _jwst2data(file_obj, ext, data_label)
jdaviz/configs/imviz/plugins/parsers.py:219: in _jwst2data
    with asdf.open(file_obj) as af:
.../asdf/asdf.py:1723: in open_asdf
    return AsdfFile._open_impl(instance,
.../asdf/asdf.py:906: in _open_impl
    generic_file = generic_io.get_file(fd, mode=mode, uri=uri)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
...
        elif mode == 'r' and hasattr(init, 'read'):
            return InputStream(init, mode, uri=uri, close=close)

>       raise ValueError("Can't handle '{0}' as a file for mode '{1}'".format(
            init, mode))
E       ValueError: Can't handle '[
<astropy.io.fits.hdu.image.PrimaryHDU object at ...>,
<astropy.io.fits.hdu.image.ImageHDU object at ...>,
<astropy.io.fits.hdu.image.ImageHDU object at ...>,
<astropy.io.fits.hdu.image.ImageHDU object at ...>,
<astropy.io.fits.hdu.image.ImageHDU object at ...>,
<astropy.io.fits.hdu.image.ImageHDU object at ...>,
<astropy.io.fits.hdu.image.ImageHDU object at ...>,
<astropy.io.fits.hdu.image.ImageHDU object at ...>,
<astropy.io.fits.hdu.table.BinTableHDU object at ...>]' as a file for mode 'r'

.../asdf/generic_io.py:1062: ValueError
```

This is the same traceback reported by CI (https://github.com/spacetelescope/jdaviz/pull/726/checks?check_run_id=3032442124)

**Comment by jdavies-st** [comment]
It looks like you're passing an open HDUList to `asdf.open()` which, it doesn't know what to do with.  @eslavich I wonder if it should know to pass an open HDUList along to `AsdfInFits.open()` instead?

Anyway, you can pass it the file or file object instead.  If you need to pass an open HDUList object, then you can do something like:

```python
from asdf.fits_embed import AsdfInFits
from astropy.io import fits


with fits.open("jw42424001001_01101_00001_nrca5_cal.fits") as hdulist:
    with AsdfInFits.open(hdulist) as af:
        wcs = af.tree["meta"]["wcs"]
        ...
```

**Comment by pllim** [comment]
Well, `jwst.datamodels` can handle that without problem. I'd rather wait for `asdf` to support such usage too. There is no rush to remove `jwst` as dependency, is there?

**Comment by jdavies-st** [comment]
`AsdfInFits.open()` does support passing an open HDUList, so that's probably the more explicit solution anyway.  And that's certainly not going away.

**Comment by jdavies-st** [comment]
And no, no rush.  =)

**Comment by pllim** [comment]
Yes, simple search and replace seemed to work. Thanks for the suggestion, @jdavies-st !

File: setup.cfg [hunk 1]
```diff
@@ -51,7 +51,8 @@ docs =
     sphinx-astropy
 all =
     scikit-image
-    jwst>=1.2.1
+    asdf
+    gwcs
```
[the remark this answers is not in the document]
Comment (pllim) [reply setup.cfg:55]: 🤷 

If you want, I can grab the pins from `jwst` if any, or just pin to latest releases.

[the remark this answers is not in the document]
Comment (ibusko) [reply setup.cfg:55]: Either way is good from my standpoint. I just wonder if we have a policy for these things, considering long-term maintenance and such. Given that for `jwst` it's a `>=` condition, it's probably OK to leave unspecified, just as is.

[the remark this answers is not in the document]
Comment (pllim) [reply setup.cfg:55]: @jdavies-st , what would be your advise since you know what feature was added when?

[the remark this answers is not in the document]
Comment (eslavich) [reply setup.cfg:55]: We're trying to stick to semantic versioning for asdf, and are planning a laundry list of breaking changes for 3.0, so a `<3` specifier might be a good idea.

[the remark this answers is not in the document]
Comment (pllim) [reply setup.cfg:55]: I am only using `AsdfInFits.open`. Are you going to break that too? I guess I can also make our "devdeps" job build asdf from source... 🤔

[the remark this answers is not in the document]
Comment (jdavies-st) [reply setup.cfg:55]: Probably `asdf>=2.7.4` would be sufficient, as that added support for streaming files over HTTP via Box.  And `gwcs>=0.16.1`, which is what jwst is currently using.

**Comment by pllim** [comment]
I added `asdf` and `gwcs` dev into the devdeps job. So far so good, but if it starts failing, you'll hear from me for sure. 😉  So, I don't think we need to pin maxversion right now.

Not sure what's up with the coverage but I wouldn't worry about it.

**Comment by pllim** [comment]
Why is it complaining that asdf/gwcs block isn't run? I need to investigate!

**Comment by pllim** [comment]
Okay, there was a bug in `asdf` import causing false positive. After I fixed the import, I find that suggestion from @jdavies-st in https://github.com/spacetelescope/jdaviz/pull/726#issuecomment-878432504 actually does not work. To try it for yourself without using `jdaviz`, see code below. We're not back to square one.

```python
>>> import asdf
>>> import gwcs
>>> from asdf.fits_embed import AsdfInFits
>>> from astropy.io import fits
>>> from astropy.utils.data import download_file
>>> asdf.__version__
'2.8.1'
>>> gwcs.__version__
'0.16.1'
>>> jwst_asdf_url_1 = 'https://data.science.stsci.edu/redirect/JWST/jwst-data_analysis_tools/stellar_photometry/jw01072001001_01101_00001_nrcb1_cal.fits'
>>> filename = download_file(jwst_asdf_url_1, cache=True)
>>> with fits.open(filename) as pf:
...     with AsdfInFits.open(pf) as af:
...         print(af.tree)
```

```
.../asdf/entry_points.py:44: AsdfWarning: asdf_extensions plugin from package stpipe==0.2.0 failed to load:

DistributionNotFound: The 'crds>=7.4.1.3' distribution was not found and is required by the application
  warnings.warn(
.../asdf/yamlutil.py:310: AsdfConversionWarning: tag:stsci.edu:jwst_pipeline/v23tosky-0.7.0 is not recognized, converting to raw Python data structure
  warnings.warn("{} is not recognized, converting to raw Python "
.../asdf/yamlutil.py:333: AsdfConversionWarning: Failed to convert tag:stsci.edu:asdf/transform/compose-1.1.0 to custom type (detail: Unknown model type 'tag:stsci.edu:jwst_pipeline/v23tosky-0.7.0'). Using raw Python data structure instead
  warnings.warn("Failed to convert {} to custom type (detail: {}). "
.../asdf/yamlutil.py:333: AsdfConversionWarning: Failed to convert tag:stsci.edu:gwcs/step-1.0.0 to custom type (detail: "transform" should be an instance of astropy.modeling.Model.). Using raw Python data structure instead
  warnings.warn("Failed to convert {} to custom type (detail: {}). "
---------------------------------------------------------------------------

.../asdf/fits_embed.py in open(cls, fd, uri, validate_checksums, extensions, ignore_version_mismatch, ignore_unrecognized_tag, strict_extension_check, ignore_missing_extensions, **kwargs)
    216             to be valid.
    217         """
--> 218         return cls._open_impl(fd, uri=uri,
    219                        validate_checksums=validate_checksums,
    220                        extensions=extensions,

.../asdf/fits_embed.py in _open_impl(cls, fd, uri, validate_checksums, extensions, ignore_version_mismatch, ignore_unrecognized_tag, strict_extension_check, ignore_missing_extensions, **kwargs)
    260
    261         try:
--> 262             return cls._open_asdf(self, generic_file,
    263                               validate_checksums=validate_checksums,
    264                               extensions=extensions,

.../asdf/asdf.py in _open_asdf(cls, self, fd, validate_checksums, extensions, _get_yaml_content, _force_raw_types, strict_extension_check, ignore_missing_extensions, **kwargs)
    884                     raise
    885
--> 886             tree = yamlutil.tagged_tree_to_custom_tree(tree, self, _force_raw_types)
    887
    888             if not (ignore_missing_extensions or _force_raw_types):

.../asdf/yamlutil.py in tagged_tree_to_custom_tree(tree, ctx, force_raw_types, _serialization_context)
    337         return node
    338
--> 339     return treeutil.walk_and_modify(
    340         tree,
    341         _walker,

.../asdf/treeutil.py in walk_and_modify(top, callback, ignore_implicit_conversion, postorder, _context)
    418
    419     with _context:
--> 420         return _recurse(top)
    421         # Generators will be drained here, if this is the outermost
    422         # call to walk_and_modify.

.../asdf/treeutil.py in _recurse(node, json_id)
    399                 # If this is a postorder modification, invoke the
    400                 # callback on this node's children first.
--> 401                 result = _handle_children(node, json_id)
    402                 result = _handle_callback(result, json_id)
    403             else:

.../asdf/treeutil.py in _handle_children(node, json_id)
    375             result = node
    376
--> 377         return _handle_generator(result)
    378
    379     def _recurse(node, json_id=None):

.../asdf/treeutil.py in _handle_generator(result)
    273         if isinstance(result, types.GeneratorType):
    274             generator = result
--> 275             result = next(generator)
    276             _context.add_generator(generator)
    277

.../asdf/treeutil.py in _handle_mapping(node, json_id)
    301                 result[key] = PendingValue
    302             else:
--> 303                 value = _recurse(value, json_id)
    304                 if value is not RemoveNode:
    305                     result[key] = value

.../asdf/treeutil.py in _recurse(node, json_id)
    399                 # If this is a postorder modification, invoke the
    400                 # callback on this node's children first.
--> 401                 result = _handle_children(node, json_id)
    402                 result = _handle_callback(result, json_id)
    403             else:

.../asdf/treeutil.py in _handle_children(node, json_id)
    375             result = node
    376
--> 377         return _handle_generator(result)
    378
    379     def _recurse(node, json_id=None):

.../asdf/treeutil.py in _handle_generator(result)
    273         if isinstance(result, types.GeneratorType):
    274             generator = result
--> 275             result = next(generator)
    276             _context.add_generator(generator)
    277

.../asdf/treeutil.py in _handle_mapping(node, json_id)
    301                 result[key] = PendingValue
    302             else:
--> 303                 value = _recurse(value, json_id)
    304                 if value is not RemoveNode:
    305                     result[key] = value

.../asdf/treeutil.py in _recurse(node, json_id)
    400                 # callback on this node's children first.
    401                 result = _handle_children(node, json_id)
--> 402                 result = _handle_callback(result, json_id)
    403             else:
    404                 # Otherwise, invoke the callback on the node first,

.../asdf/treeutil.py in _handle_callback(node, json_id)
    280     def _handle_callback(node, json_id):
    281         if callback_arity == 1:
--> 282             result = callback(node)
    283         else:
    284             result = callback(node, json_id)

.../asdf/yamlutil.py in _walker(node)
    329         # not.
    330         try:
--> 331             return tag_type.from_tree_tagged(node, ctx)
    332         except TypeError as err:
    333             warnings.warn("Failed to convert {} to custom type (detail: {}). "

.../asdf/types.py in from_tree_tagged(cls, tree, ctx)
    358             An instance of the custom type represented by this extension class.
    359         """
--> 360         return cls.from_tree(tree.data, ctx)
    361
    362     @classmethod

.../gwcs/tags/wcs.py in from_tree(cls, node, ctx)
     27     def from_tree(cls, node, ctx):
     28         name = node['name']
---> 29         steps = [(x.frame, x.transform) for x in node['steps']]
     30         return WCS(steps, name=name)
     31

.../gwcs/tags/wcs.py in <listcomp>(.0)
     27     def from_tree(cls, node, ctx):
     28         name = node['name']
---> 29         steps = [(x.frame, x.transform) for x in node['steps']]
     30         return WCS(steps, name=name)
     31

AttributeError: 'TaggedDict' object has no attribute 'frame'
```

**Review by ibusko — APPROVED** [review]

**Comment by jdavies-st** [comment]
The reason it is not working is that the test data is old:

```
In [1]: from astropy.io import fits

In [2]: hdulist = fits.open("https://data.science.stsci.edu/redirect/JWST/jwst-data_analysis_tools/stellar_photometry/jw01072001001_01101_00001_nrcb1_cal.fits")

In [4]: print(hdulist["PRIMARY"].header["CAL_VER"])
0.14.2
```

This is very, very old data that has not been run through the `jwst` pipeline anytime in the last year.  That release is from Nov 2019 to be specific, and lots has changed since then, specifically the removal of dependence on specific jwst transforms.

Use data that has been run through `jwst 1.1` or later.

**Comment by pllim** [comment]
@eteq , I believe you provided https://data.science.stsci.edu/redirect/JWST/jwst-data_analysis_tools/stellar_photometry/jw01072001001_01101_00001_nrcb1_cal.fits for me to use. Is it possible to update that file on Box?

We shouldn't be distributing stale data via data.science.stsci.edu as "official" data for "jwst-data_analysis_tools"?

File: jdaviz/configs/imviz/tests/test_parser.py [hunk 2]
```diff
@@ -58,7 +58,7 @@ def test_validate_bunit():
 
 class TestParseImage:
     def setup_class(self):
-        self.jwst_asdf_url_1 = 'https://data.science.stsci.edu/redirect/JWST/jwst-data_analysis_tools/stellar_photometry/jw01072001001_01101_00001_nrcb1_cal.fits'  # noqa: E501
+        self.jwst_asdf_url_1 = 'https://data.science.stsci.edu/redirect/JWST/jwst-data_analysis_tools/imviz_test_data/jw00042001001_01101_00001_nrcb5_cal.fits'  # noqa: E501
```
Comment (pllim) [inline jdaviz/configs/imviz/tests/test_parser.py:61]: Many thanks to @orifox for providing a valid data file. This PR is now ready!

**Review by rosteen — APPROVED** [review]
I confirmed that the test JWST file loads in the app, approved.

**Comment by camipacifici** [comment]
JWST images are loaded fine without installing the `jwst` package. So this PR technically works. The images I used for testing are JWST simulated direct images with populated WCS ([image1](https://stsci.box.com/s/ox91bao6ijfqymqebkg6dmugurc0n5vr) and [image2](https://stsci.box.com/s/0c6srs550821hw0x3c8ew6bd9vdttmgf)). 

When switching between the images however, one of them goes blank, although the values of `pixel`, `value`, and `world` seem correct. @pllim I might have seen something about this in another ticket, but I cannot find it now. If it has been reported, then I approve this one.

**Comment by pllim** [comment]
Hm.... if they are linked, should not go blank. I'll double check.

**Comment by pllim** [comment]
@camipacifici , I cannot even load your data. I get `OSError: No SIMPLE card found, this file does not appear to be a valid FITS file`

**Comment by pllim** [comment]
Update: Turns out the URLs do not route to the actual file. If I manually download them, I don't get the FITS error.

**Comment by pllim** [comment]
> When switching between the images however, one of them goes blank

I cannot reproduce this problem. Blank works fine for me.

**Comment by pllim** [comment]
p.s. Awww nooo force push mishap. I will fix this.

**Comment by camipacifici** [comment]
Sorry about the non fits links. I picked the wrong one from box, apologies.

**Comment by pllim** [comment]
p.s. This might conflict with #778 . I'll deal with the conflict accordingly, depending on which one gets merged first.

**Review by duytnguyendtn — APPROVED** [review]
As with @rosteen, the supplied JWST file loads fine, so I'll be approving.

As a breadcrumb, I got the following traceback when trying to retrieve regions, but I don't think it's related to this PR.
```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
~\AppData\Local\Temp/ipykernel_35360/3606753409.py in <module>
----> 1 regions = imviz.get_interactive_regions()

\env\lib\site-packages\jdaviz\configs\imviz\helper.py in get_interactive_regions(self)
    509 
    510         """
--> 511         return self.app.get_subsets_from_viewer('viewer-1')
    512 
    513 

\env\lib\site-packages\jdaviz\app.py in get_subsets_from_viewer(self, viewer_reference, data_label, subset_type)
    488         """
    489         viewer = self.get_viewer(viewer_reference)
--> 490         data = self.get_data_from_viewer(viewer_reference,
    491                                          data_label,
    492                                          cls=None)

\env\lib\site-packages\jdaviz\app.py in get_data_from_viewer(self, viewer_reference, data_label, cls, include_subsets)
    431                     #  output data type
    432                     elif len(layer_data.shape) == 2:
--> 433                         layer_data = layer_data.get_object(cls=CCDData)
    434 
    435                     data[label] = layer_data

\env\lib\site-packages\glue\core\data.py in get_object(self, cls, **kwargs)
    287         handler, _ = data_translator.get_handler_for(cls)
    288 
--> 289         return handler.to_object(self, **kwargs)
    290 
    291     @property

\env\lib\site-packages\glue_astronomy\translators\ccddata.py in to_object(self, data_or_subset, attribute)
     42             wcs = None
     43         else:
---> 44             raise TypeError('data.coords should be an instance of Coordinates or WCS')
     45 
     46         if isinstance(attribute, str):

TypeError: data.coords should be an instance of Coordinates or WCS
```

**Comment by pllim** [comment]
> TypeError: data.coords should be an instance of Coordinates or WCS

That is #781 , with #786 waiting for review. 😉

**Comment by duytnguyendtn** [comment]
Ahh didn't realize it was related to `GWCS`. Good to know!

**Pull request merged**